src/System.Threading.Thread/src/System/Threading/Thread.cs

```diff
@@ -190,14 +191,44 @@ public static IPrincipal CurrentPrincipal
         {
             get
             {
-                return CurrentThread._principal ?? (CurrentThread._principal = AppDomain.CurrentDomain.GetThreadPrincipal());
+                if (CurrentThread._principal is null)
+                {
+                    IPrincipal principal = AppDomain.CurrentDomain.GetThreadPrincipal();
+                    if (principal is null)
```

Comment:
I am wondering whether these null shortcuts can cause problems. I cannot think about anything, just pointing it out.

Reply:
On issue you specified to be "lazy" on AsyncLocal instancing https://github.com/dotnet/corefx/issues/34489#issuecomment-453171829 and allocate only on set, so my idea is to avoid allocation for "default" .net core policy get, I don't see particular issue to return null, I think will be used to compare with something else.
BTW you "feeling" take precedence over my simple usage expectations so if you're more confortable with simple assignment to `CurrentPrincipal` I won't oppose resistance :smiley:

Reply:
Tried different way, similar to previous, should be lazy as possible.

Reply:
I was concerned about behavior differences between .NET Core and .NET Framework. The .NET Framework implementation does not have a null check like this and that is what made me think about this. The restate my question: Are there cases where this null check is going cause observable behavior differences between .NET Framework and .NET Core? (As I have said, I cannot think about any.)

Wrt. lazy: My primary concern was about creating the AsyncLocal only once somebody calls the CurrentPrincipal property.

Reply:
Now the only shortcut is for set `null` without previous set something, do you feel ok with it?

Reply:
Yes, I think it is ok based on the codereview. I assume that you have done enough testing to convince yourself that it behaves the same way as .NET Framework.